Initial stab at a TypeScript blog post (#5101)
sveltejs/svelte: site/content/blog/2020-06-04-svelte-and-typescript.md

---
title: Svelte <3 TypeScript
description: Typernetically enhanced web apps

> author: Orta Therox

Reviewer: You have a truly great name.

authorURL: https://twitter.com/orta
---

> It's been on the TODO list for a while, and it's now happening. TypeScript with Svelte is now a first class citizen of the eco-system.

Reviewer 2: nit: it's generally just "ecosystem"

We think it'll help you handle much larger Svelte code bases regardless of whether you use TypeScript.

## What does it mean to support TypeScript in Svelte?

TypeScript support in Svelte has been possible for a long time, but you had to mix a lot of disparate tools together and each project ran independently. Today, nearly all of these tools live under the Svelte organization and are maintained by a set of people who take responsibility over the whole pipeline and have common goals.

A week before COVID was declared a pandemic, [I pitched a consolidation](https://github.com/sveltejs/svelte/issues/4518) of the best Svelte tools and ideas from similar dev-ecosystems and provided a set of steps to get TypeScript support first class. Since then, many people have pitched in and wrote the code to get us there.

#### How does it work?

> To understand the two main parts of TypeScript support, we'll compare it to the technique TypeScript uses to provide dev tools. There is a compiler `tsc` which you run on the command-line to convert `*.ts` to `*.js`, then there is a `TSServer` which is a node API that responds to requests from text editors. The `TSServer` is what provides all the JavaScript and TypeScript experience, and it has most of the compiler's code inside it.

Reviewer 3: `what provides all the JavaScript and TypeScript experience` -> I would rephrase it to `what provides all the JavaScript and TypeScript coding experience` (`coding` was added), so that it's more clear that we talk about IDE support.

Author: I like the idea, I think I might angle it more at "realtime information"

For Svelte, we have the Svelte compiler, and now we have the [`svelte-language-server`](https://github.com/sveltejs/language-tools/tree/master/packages/language-server#svelte-language-server) which responds to text editor calls via the [Language Server Protocol standard](https://microsoft.github.io//language-server-protocol/overviews/lsp/overview/). First class TypeScript support means that _both_ of these two systems do good job of handling TypeScript code.

The Svelte compiler support for TypeScript is handled by [Christian Kaisermann](https://github.com/kaisermann)'s [`svelte-preprocess`](https://github.com/sveltejs/svelte-preprocess#svelte-preprocess) which is now an official Svelte project.

> For the editor level, we took inspiration from [Pine's](https://github.com/octref) work in the [Vue](https://vuejs.org) ecosystem via [Vetur](https://github.com/vuejs/vetur). Vetur provides an [LSP](https://github.com/vuejs/vetur/blob/master/server), a VS Code extension and a [CLI](https://github.com/vuejs/vetur/blob/master/vti). Svelte now also has an [LSP](https://github.com/sveltejs/language-tools/blob/master/packages/language-server), a [VS Code extension](https://github.com/sveltejs/language-tools/blob/master/packages/svelte-vscode) and a [CLI](https://github.com/sveltejs/language-tools/blob/master/packages/svelte-check).

Reviewer 3: Not a native speaker, but isn't i `a LSP` instead of `an LSP`?

Author: I ended up duckduckgoing the phrase, and it looks like "an LSP" is generally what everyone seems to write - so sticking with that 👍

Reviewer 2: Yeah, "an LSP" is right. The reason why is that you say it like "el es pee" and so it starts with a vowel sound

Reviewer: Yeah, the choice of article is always based on the phonetic quality. Which is very confusing for written words.

#### `*.svelte` Introspection

> For the official Svelte VS Code extension, we built off the foundations which [James Britles](https://github.com/UnwrittenFun) has created in [`UnwrittenFun/svelte-vscode`](https://github.com/UnwrittenFun/svelte-vscode) and [`UnwrittenFun/svelte-language-server`](https://github.com/UnwrittenFun/svelte-language-server/).

Reviewer: Typo on the name (I think):

```suggestion
For the official Svelte VS Code extension, we built off the foundations which [James Birtles](https://github.com/UnwrittenFun) has created in [`UnwrittenFun/svelte-vscode`](https://github.com/UnwrittenFun/svelte-vscode) and [`UnwrittenFun/svelte-language-server`](https://github.com/UnwrittenFun/svelte-language-server/).
```

> [Simon H](https://github.com/dummdidumm) and [Lyu, Wei-Da](https://github.com/jasonlyu123) have worked on improving JavaScript and typeScript improvements, including integrating [@halfnelson](https://github.com/halfnelson)'s [svelte2ts](https://github.com/sveltejs/language-tools/tree/master/packages/svelte2tsx#svelte2tsx) which powers understanding the props on components in your codebase.

Reviewer 4: small nit: typescript --> TypeScript. This blog post is great.

> [Simon H](https://github.com/dummdidumm) and [Lyu, Wei-Da](https://github.com/jasonlyu123) have worked on improving JavaScript and typeScript improvements, including integrating [@halfnelson](https://github.com/halfnelson)'s [svelte2ts](https://github.com/sveltejs/language-tools/tree/master/packages/svelte2tsx#svelte2tsx) which powers understanding the props on components in your codebase.

Reviewer 5: svelte2ts should probably be svelte2tsx
```suggestion
[Simon H](https://github.com/dummdidumm) and [Lyu, Wei-Da](https://github.com/jasonlyu123) have worked on improving JavaScript and typeScript improvements, including integrating [@halfnelson](https://github.com/halfnelson)'s [svelte2tsx](https://github.com/sveltejs/language-tools/tree/master/packages/svelte2tsx#svelte2tsx) which powers understanding the props on components in your codebase.
```

> [Simon H](https://github.com/dummdidumm) and [Lyu, Wei-Da](https://github.com/jasonlyu123) have worked on improving JavaScript and typeScript improvements, including integrating [@halfnelson](https://github.com/halfnelson)'s [svelte2ts](https://github.com/sveltejs/language-tools/tree/master/packages/svelte2tsx#svelte2tsx) which powers understanding the props on components in your codebase.

Reviewer 5: Having the words improvement and improving here seems redundant maybe something like?
```suggestion
[Simon H](https://github.com/dummdidumm) and [Lyu, Wei-Da](https://github.com/jasonlyu123) have done great work improving the Javascript and Typescript introspection, including integrating [@halfnelson](https://github.com/halfnelson)'s [svelte2ts](https://github.com/sveltejs/language-tools/tree/master/packages/svelte2tsx#svelte2tsx) which powers understanding the props on components in your codebase.
```

Reviewer 6: Was about to suggest the same. Also typescript should be TypeScript. Great post, really like the wording!

> [Simon H](https://github.com/dummdidumm) and [Lyu, Wei-Da](https://github.com/jasonlyu123) have worked on improving JavaScript and typeScript improvements, including integrating [@halfnelson](https://github.com/halfnelson)'s [svelte2ts](https://github.com/sveltejs/language-tools/tree/master/packages/svelte2tsx#svelte2tsx) which powers understanding the props on components in your codebase.

Reviewer 3: You can write my full name here: Simon Holthausen.

### Try today

The Svelte template has been extended with a script to convert it to a TypeScript project.

```bash

> npx degit svelte/template

Reviewer 6: I think it should be `npx degit sveltejs/template`

Reviewer: I think you are looking at an old diff, this has been changed.

> node scripts/updateTypeScriptVersion.js

Reviewer 7: I really think this should be turned into a `package.json` task - like `npm run typescript:init` or something. That script name is quite wordy.

Reviewer: That should be discussed over at : https://github.com/sveltejs/template/pull/140

Reviewer 7: I feel like I already mentioned it on an existing PR. I don't recognise this PR.

Author: I don't have _too_ strong an opinion on it, but I lean towards using the script directly. The key advantage is that it's 1 command to remove it ( `rm scripts/updateTypeScriptVersion.js`) if you're don't want TS and want to carry on in JS

if we add the script in the package JSON, then it's `rm ...` the file then edit the package manually.

Reviewer 6: I agree with @antony, I think a `npm run ts:init` would be more comfortable, but I do understand @orta 's considerations about it.
Perhaps we could have some `svelte-cli` or similar package that would allow us to do things like `npx svelte-cli ts:init` and also `npx svelte-cli ts:check` and so. I thinks it's a clean solution.
On the other hand, I think that we should also support something like `npx degit sveltejs/template#ts my-svelte-ts-project` to start a new project with typescript support. That would defintely be the easiest way to get started without having to mess with the conf files.

Reviewer 8: I think at this point we should probably have an `npm init svelte` command* with an interactive 'TypeScript? y/n' prompt (and possibly others). But that's a larger conversation — for now let's roll with `node scripts/setupTypeScript.js`

*this would require us to have an npm package called `create-svelte`, which [already exists](https://www.npmjs.com/package/create-svelte), so we'd need to see if the owner would consider transferring it. or we could have `@sveltejs/create` which you'd use as `npm init @sveltejs`

```

### Adding TypeScript support to your project

##### 1. Compiling TypeScript

You first need to set up [`svelte-preprocess`](https://github.com/sveltejs/svelte-preprocess#svelte-preprocess) and  [`@rollup/plugin-typescript`](https://github.com/rollup/plugins/tree/master/packages/typescript#rollupplugin-typescript) for the compiler:

When in a rollup project, this would mean:

> ```diff

Reviewer 7: didn't know you could have a lang of diff. amazing!

+ import preprocess from 'svelte-preprocess'
+ import typescript from '@rollup/plugin-typescript';

export default {
  ...,
  plugins: [
    svelte({
+       preprocess: autoPreprocess({ /* options */ })
    }).
+   typescript({ sourceMap: !production }),
  ]
}
```

[Full instructions for other environments here](https://github.com/sveltejs/svelte-preprocess#usage)

To configure TypeScript, you will need to create a `tsconfig.json` in the root of your project:

```json
{
  "extends": "@tsconfig/svelte/tsconfig.json",

  "include": ["src/**/*"],
  "exclude": ["node_modules/*", "__sapper__/*", "public/*"],
}
```

Your `include`/`exclude` may differ per project.

##### 2. Editor Support

Any editor which supports using an LSP can work, the [VS Code](https://marketplace.visualstudio.com/items?itemName=svelte.svelte-vscode) extension has been our primary focus but there is work in progress [on Atom](https://github.com/sveltejs/language-tools/pull/160), and [coc-svelte](https://github.com/coc-extensions/coc-svelte) has been updated with the latest changes.

> These editor extensions will work with your existing JavaScript code. The editor won't offer errors, but it will offer inference and refactoring tools. You can [add `// @check-js`](https://www.staging-typescript.org/docs/handbook/intro-to-js-ts.html) to the top of a `<script>` tag using JavaScript to get better error messages.

Reviewer 3: Rephrasing a little here to make it more clear that the extension is a must, even if you don't use TypeScript 😄 

```suggestion
These editor extensions will improve your coding experience even if you only use JavaScript. The editor won't offer errors, but it will offer inference and refactoring tools. You can [add `// @check-js`](https://www.staging-typescript.org/docs/handbook/intro-to-js-ts.html) to the top of a `<script>` tag using JavaScript to get better error messages.
```

#### `*.svelte` Introspection

For the official Svelte VS Code extension, we built off the foundations which [James Birtles](https://github.com/UnwrittenFun) has created in [`UnwrittenFun/svelte-vscode`](https://github.com/UnwrittenFun/svelte-vscode) and [`UnwrittenFun/svelte-language-server`](https://github.com/UnwrittenFun/svelte-language-server/).

[Simon Holthausen](https://github.com/dummdidumm) and [Lyu, Wei-Da](https://github.com/jasonlyu123) have done great work improving the JavaScript and TypeScript introspection, including integrating [@halfnelson](https://github.com/halfnelson)'s [svelte2tsx](https://github.com/sveltejs/language-tools/tree/master/packages/svelte2tsx#svelte2tsx) which powers understanding the props on components in your codebase.

### Try today

The Svelte template has been extended with a script to convert it to a TypeScript project.

```bash

> npx degit svelte/template

Reviewer 9: ```
$ npx degit svelte/template
! could not find commit hash for master
```

Shouldn't it be `npx degit sveltejs/template svelte-app` ? (ref : https://github.com/sveltejs/template#svelte-app)

---
title: Svelte <3 TypeScript
description: Typernetically enhanced web apps
author: Orta Therox
authorURL: https://twitter.com/orta
---

> It's been on the TODO list for a while, and it's now happening. TypeScript with Svelte is now a first class citizen of the ecosystem.

Reviewer 8: I think we're underselling how big a deal this is!

```suggestion
It's been by far the most requested feature for a while, and it's finally here: Svelte officially supports TypeScript.
```

> We think it'll help you handle much larger Svelte code bases regardless of whether you use TypeScript or JavaScript.

Reviewer 8: Might be better if this doesn't suggest that it's only applicable to large codebases? We don't want people who use Svelte for smaller projects to feel like our focus is shifting away from them, since it isn't

```suggestion
We think it'll give you a much nicer development experience — one that also scales beautifully to larger Svelte code bases — regardless of whether you use TypeScript or JavaScript.
```

<figure>
	<img alt="Screenshot of TypeScript in Svelte" src="media/svelte-ts.png">
	<figcaption>Image of TypeScript + Svelte in VS Code (theme is <a href="https://marketplace.visualstudio.com/items?itemName=karyfoundation.theme-karyfoundation-themes">Kary Pro</a>.)</figcaption>
</figure>

> 

Reviewer 8: I think a lot of people will want to just try it out straight away, so here (above the fold) would be a good place to have the quickstart guide

```suggestion

## Try it now

You can start a new Svelte TypeScript project using the [normal template](https://github.com/sveltejs/template) by running `node scripts/setupTypeScript.js` before you do anything else:

    npx degit sveltejs/template svelte-typescript-app
    cd svelte-typescript-app
    node scripts/setupTypeScript.js
    
If you're a VS Code user, make sure you're using the (new) [official extension](https://marketplace.visualstudio.com/items?itemName=svelte.svelte-vscode), which replaces the popular extension by James Birtles.

Later in this blog post, we'll detail the individual steps involved in using TypeScript in an existing Svelte project.

```

(side-note: how do you put code blocks inside a commit suggestion...?)

Reviewer: Possibly by giving the `suggestion` language 4 backticks instead of 3. But don't quote me on this.

Author: This week I learned how:

GFMarkdown allows for the code fences to be an arbitrary size

````
```ts

```
````

## What does it mean to support TypeScript in Svelte?

> 

Reviewer 8: I think this section should first answer the question 'what does TS support mean in practice?':

```suggestion

When we say that Svelte now supports TypeScript, we mean a few different things:

* You can use TypeScript inside your `<script>` blocks — just add the `lang="ts"` attribute
* Components with TypeScript can be typechecked with the `svelte-check` command
* You get autocompletion hints and typechecking as you're writing components, even in expressions inside markup
* TypeScript files understand the Svelte component API — no more red squiggles when you import a `.svelte` file into a `.ts` module

```

Author: Agree, I think this section adds 👍

TypeScript support in Svelte has been possible for a long time, but you had to mix a lot of disparate tools together and each project ran independently. Today, nearly all of these tools live under the Svelte organization and are maintained by a set of people who take responsibility over the whole pipeline and have common goals.

> A week before COVID was declared a pandemic, [I pitched a consolidation](https://github.com/sveltejs/svelte/issues/4518) of the best Svelte tools and ideas from similar dev-ecosystems and provided a set of steps to get TypeScript support first class. Since then, many people have pitched in and wrote the code to get us there.

Reviewer 8: minor grammar tweak

```suggestion
A week before COVID was declared a pandemic, [I pitched a consolidation](https://github.com/sveltejs/svelte/issues/4518) of the best Svelte tools and ideas from similar dev-ecosystems and provided a set of steps to get first class TypeScript support. Since then, many people have pitched in and written the code to get us there.
```

#### `*.svelte` Introspection

For the official Svelte VS Code extension, we built off the foundations which [James Birtles](https://github.com/UnwrittenFun) has created in [`UnwrittenFun/svelte-vscode`](https://github.com/UnwrittenFun/svelte-vscode) and [`UnwrittenFun/svelte-language-server`](https://github.com/UnwrittenFun/svelte-language-server/).

[Simon Holthausen](https://github.com/dummdidumm) and [Lyu, Wei-Da](https://github.com/jasonlyu123) have done great work improving the JavaScript and TypeScript introspection, including integrating [@halfnelson](https://github.com/halfnelson)'s [svelte2tsx](https://github.com/sveltejs/language-tools/tree/master/packages/svelte2tsx#svelte2tsx) which powers understanding the props on components in your codebase.

> ### Try today

Reviewer 8: As mentioned above I think this should go near the top of the blog post, so we could maybe remove it from down here?

Author: Deleted it, now that there's a try above the fold

The Svelte template has been extended with a script to convert it to a TypeScript project.

```bash
npx degit sveltejs/template svelte-app
cd svelte-app
node scripts/updateTypeScriptVersion.js
```

> ### Adding TypeScript support to your project

Reviewer 8: ```suggestion
### Adding TypeScript support to an existing project
```

Before getting started, add the dependencies:

```bash

> yarn add --dev svelte-preprocess @rollup/plugin-typescript @tsconfig/svelte svelte-check typescript

Reviewer 8: I think we should just have instructions for `npm` — at first glance my instinct was to copy and paste both lines into the terminal. People who use `yarn` already understand how to translate `npm` commands, I don't think we don't need to be explicit

> npm install --save-dev svelte-preprocess @rollup/plugin-typescript @tsconfig/svelte svelte-check typescript

Reviewer 8: Not 100% on this, but perhaps we should leave out mention of `@rollup/plugin-typescript` here, since it's not Svelte-specific and it sort of gives the impression that Rollup is a prerequisite?

```suggestion
npm install --save-dev @tsconfig/svelte typescript svelte-preprocess svelte-check 
```

Author: I think the split feels meaningful

```

##### 1. Compiling TypeScript

> You first need to set up [`svelte-preprocess`](https://github.com/sveltejs/svelte-preprocess#svelte-preprocess) and  [`@rollup/plugin-typescript`](https://github.com/rollup/plugins/tree/master/packages/typescript#rollupplugin-typescript) for the compiler:

Reviewer 8: ```suggestion
You first need to set up [`svelte-preprocess`](https://github.com/sveltejs/svelte-preprocess#svelte-preprocess), which passes the contents of your `<script lang="ts">` blocks through the TypeScript compiler.
```

> When in a rollup project, this would mean:

Reviewer 8: ```suggestion
In a Rollup project, that would look like this — note that we also need to install `@rollup/plugin-typescript` so that Rollup can handle `.ts` files:
```

```diff

> + import autoPreprocess from 'svelte-preprocess'

Reviewer 10: Nit: Append `;` for consistency with the next line and the default in `rollup.config.js`

Reviewer 10: Also just realized the script `updateTypeScriptVersion.js` loads this as `sveltePreprocess`, not `autoPreprocess`.

> + import autoPreprocess from 'svelte-preprocess'

Reviewer 8: ```suggestion
+ import sveltePreprocess from 'svelte-preprocess';
```

Author: This is to keep it inline with the svelte-preprocess docs: https://github.com/sveltejs/svelte-preprocess#what-is-it

![Screen Shot 2020-07-17 at 5 05 51 PM](https://user-images.githubusercontent.com/49038/87830770-c61d4500-c84f-11ea-87a2-113ca70cb1ff.png)

Reviewer 11: I'm open to changing that 😁  I've already named the default export `preprocess`, `autoPreprocess`, and `sveltePreprocess`. My only concern is blurring the line between the `preprocess` property and `svelte-preprocess` even more. I also thought about exporting an `auto` preprocessor instead of using the default export (keeping the default for compatibility, of course).

Author: I don't have a strong opinion other than consistency 👍

Reviewer 2: `sveltePreprocess` makes more sense to me

Reviewer 11: Will change in `svelte-preprocess` then!

+ import typescript from '@rollup/plugin-typescript';

export default {
  ...,
  plugins: [
    svelte({

> +       preprocess: autoPreprocess({ /* options */ })

Reviewer 8: ```suggestion
+       preprocess: sveltePreprocess()
```

    }).
+   typescript({ sourceMap: !production }),
  ]
}
```

[Full instructions for other environments here](https://github.com/sveltejs/svelte-preprocess#usage)

To configure TypeScript, you will need to create a `tsconfig.json` in the root of your project:

```json
{
  "extends": "@tsconfig/svelte/tsconfig.json",

>   "include": ["src/**/*"],
>   "exclude": ["node_modules/*", "__sapper__/*", "public/*"],

Reviewer 11: For what I tested, I think this should be something like:

```suggestion
  "include": ["src/", "src/node_modules"],
  "exclude": ["node_modules/*", "__sapper__/*", "public/*"]
```

Even if we don't explicitly exclude the `node_modules/*, `src/node_modules` end up being ignored 🤷

Author: Thanks, added!

}
```

> Your `include`/`exclude` may differ per project.

Reviewer 8: ```suggestion
Your `include`/`exclude` may differ per project — these are defaults that should work across most Svelte projects.
```

Would this exclude files in `src/node_modules` in a Sapper project?

Author: `src/node_modules` - that's a wild path, but no - I don't think it would

@benmccann can you confirm, I figure you've already tried this

Reviewer 2: I've done limited testing since I'm still waiting for https://github.com/sveltejs/sapper/pull/1222 to be reviewed. @babichjacob has actually probably tested more than I have. I've been using his templates for testing. Here's what he has: https://github.com/babichjacob/sapper-typescript-graphql-template/blob/5a78283059953749acfbb8d60430187646a58212/tsconfig.json#L16

These editor extensions will improve your coding experience even if you only use JavaScript. The editor won't offer errors, but it will offer inference and refactoring tools. You can [add `// @check-js`](https://www.staging-typescript.org/docs/handbook/intro-to-js-ts.html) to the top of a `<script>` tag using JavaScript to get better error messages with no infra changes.

To switch a `<script>` to use TypeScript, use `<script lang="ts">` and that should be it. Hopefully you won't be seeing an ocean of red squiggles.

##### 3. CI Checks

Having red squiggles is great, well, kinda. On the long run though, you want to be able to verify that there are no errors in your code. To verify your project is error free, you can use the CLI tool [`svelte-check`](https://www.npmjs.com/package/svelte-check). It acts like an editor asking for errors against all of your `.svelte` files.

You can add the dependency to your project and then add it to CI.

```bash

> ❯ yarn svelte-check

Reviewer 8: re earlier suggestion about defaulting to `npm` rather than `yarn`

```suggestion
❯ npx svelte-check
```

  ...,
  plugins: [
    svelte({

> +       preprocess: autoPreprocess({ /* options */ });

Reviewer 11: ```suggestion
+       preprocess: autoPreprocess({ /* options */ }),
```

👀

Author: great point
